flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java

```diff
@@ -220,8 +221,13 @@ public void discardOnFailedStoring() throws Exception {
         discard();
     }
 
+    /**
+     * NOT Thread safe. This method can be called only from {@link CheckpointCoordinator}(under the
+     * lock).
+     */
```

Comment:
What's the threading model of this class at the moment? It looks to me as all methods except of `discard()` are called from a "single thread" (single thread = single thread or multiple threads but protected by an external lock)?

If so, maybe it makes sense to consider a construct like this:

```
@NotThreadSafe
public class CompletedCheckpoint {
  
  public DiscardObject markAsDiscarded();
  
  @NotThreadSafe
  public static class DiscardObject {
    (...);
     public void discard();
  }
}
```

Where both `CompletedCheckpoint` and `DiscardObject` are not thread safe on their own, however the `DiscardObject` can be created in one thread, passed to another, and then executed in that other thread? Kind of similar to `BufferBuilder` and `BufferConsumer`, each not thread safe individually, but could reside in two different threads.